# Tentang Fungsi
Apabila program yang dibuat sedemikian panjangnya sehingga sangat sulit untuk dimengerti dan dipahami jalannya program, maka akan lebih baik jika kita memecah program tersebut menjadi suatu program-program kecil yang biasa disebut dengan fungsi, prosedur, atau ada pula yang menyebutnya dengan modul.

Beberapa hal tentang keuntungan menggunakan fungsi :
* Pemakaian Fungsi akan mencegah penulisan kode yang berulang-ulang dalam suatu program.
* Pemakaian Fungsi memungkinkan pemrogram untuk menyusun skrip ke dalam bentuk modul-modul kecil yang lebih efisien dengan tugasnya masing-masing.

# Mendefinisikan Fungsi
Fungsi pada Python, dibuat dengan kata kunci **def** kemudian diikuti dengan nama fungsinya.

In [1]:
# Contoh
# Membuat Fungsi dengan argumen
def luaspersegi(panjang, lebar):
    luas = panjang * lebar
    print("Luas Persegi panjang =", luas)

# Memanggil Fungsi
Cara memanggil fungsi tersebut adalah dengan mengetiikan nama fungsi berikut parameternya.

Fungsi juga dapat dipanggil dan diproses di dalam fungsi lain. Fungsi juga bisa memanggil dirinya sendiri. Fungsi yang memanggil dirinya sendiri disebut dengan fungsi *rekursif.*

## Memanggil Fungsi Sederhana

In [2]:
# Contoh
# Membuat Fungsi
def halo():
    print("Halo Dunia, Selamat bertemu")
    print("Salam dari Salatiga")
    return

# Pemanggilan Fungsi
halo()

Halo Dunia, Selamat bertemu
Salam dari Salatiga


## Memanggil Fungsi dengan Argumen
Argumen adalah input dari luar yang akan dibawa dan selanjutnya diproses di dalam fungsi.

In [3]:
# Contoh
# Membuat fungsi dengan parameter
def luaspersegipanjang(panjang, lebar):
    luas = panjang * lebar
    print("Luas Persegi Panjang : %f" %luas)
    
# Pemanggilan fungsi
luaspersegipanjang(4, 6)

Luas Persegi Panjang : 24.000000


## Memanggil Fungsi yang Mengembalikan Nilai
Sebuah fungsi dapat mengembalikan nilai dari hasil pemrosesannya, sehingga dapat digunakan pada proses berikutnya. Fungsi tersebut dapat mengembalikan nilai dengan menggunakan kata kunci *return* kemudian diikuti dengan nilai atau variabel yang akan dikembalikan tersebut.

In [4]:
# Contoh
def luaspersegipanjang(panjang, lebar):
    luas = panjang * lebar
    return luas

# Pemanggilan fungsi
print("Luas : %d" %luaspersegipanjang(6, 3))

Luas : 18


# Fungsi Rekursif
Fungsi Rekursif adalah fungsi yang memanggil dirinya sendiri, sehingga terjadilah perulangan di dalam fungsi tersebut. Proses perulangan tersebut terjadi secara implisit, tidak menggunakan *statement* perulangan pada umumnya seperti *for, while,* dan sebagainya.

Fungsi rekursif memiliki kelebihan program lebih singkat dan tentunya lebih efisien dan proses yang cepat dibandingkan proses iteratif pada umumnya. Sedangkan kekurangannya adalah menggunakan memori yang lebih besar karena setiap memanggil bagian dari dirinya membutuhkan ruang memori untuk penyimpanan. Rekursif yang kurang tepat seringkali tidak bisa berhenti sehingga menghabiskan banyak memori yang terbuang percuma.

In [5]:
# Contoh fungsi faktorial
def faktorial(n):
    if n <= 1:
        return 1
    else:
        return n * faktorial(n-1)

# Program Utama
i = 0
n = int(input("Berapa Faktorial :"))

while i<=n:
    print(i,'!=',faktorial(i))
    i+=1

Berapa Faktorial :5
0 != 1
1 != 1
2 != 2
3 != 6
4 != 24
5 != 120


0 != 1<br>
1 != 1<br>
2 != 2x1 = 2<br>
3 != 3x2x1 = 6<br>
4 != 4x3x2x1 = 24<br>
5 != 5x4x3x2x1 = 120

# Tentang Scope sebuah Variabel
Ruang lingkup sebuah variabel bisa bersifat *local* atau *global.* Jadi tidak semua variabel bisa diakses dari semua tempat, tergantung bagaimana mendefinisikan variabel tersebut.

* Variabel lokal<br>
Variabel lokal adalah variabel yang hanya bisa di akses di dalam sebuah fungsi saja, dimana dia didefinisikan.

* Variabel Global<br>
Variabel Global didefinisi di luar fungsi sehingga memiliki *scope* global dan bisa diakses dari mana saja.

In [6]:
# Contoh
# Membuat variabel global
nama = "Alvin Agusta"
alamat = "Amerika Serikat"

def testing():
    # ini variabel lokal
    nama = "Navida Wahyu"
    alamat = "Salatiga"
    # mengakses variabel lokal
    print("Nama: %s" % nama)
    print("Alamat: %s"% alamat)

# Mengakses variabel global
print("VARIABEL GLOBAL")
print("Nama: %s" % nama)
print("Alamat: %s"% alamat)

# Mengakses fungsi testinglokal()
print("VARIABEL LOKAL")
testing()

VARIABEL GLOBAL
Nama: Alvin Agusta
Alamat: Amerika Serikat
VARIABEL LOKAL
Nama: Navida Wahyu
Alamat: Salatiga


Pada skrip di atas, variabel nama dan alamat yang ditulis pada baris awal adalah variabel global yang bisa dipanggil dari mana saja.<br>
Sedangkan variabel nama dan alamt yang diletakkan di dalam fungsi **testing()** adalah variabel lokal dan hanya bisa dipanggil dalam fungsi itu.